### PR TITLE
Add task to release for Mac App Store

### PR DESCRIPTION
**Pull request by rodrigok**
```shell
npm run release -- --mas --sign "3rd Party Mac Developer Application: Company Name (APPIDENTITY)" --installer-sign "3rd Party Mac Developer Installer: Konecty Company Name (APPIDENTITY)"
```

### CODE REVIEW

**Comment by engelgabriel** [comment]
+1

**Comment by szwacz** [comment]
@rodrigok wow! This is something for sure people want.
Two issues thought:
- Why you decided to do completely separate release script for OSX, can't we embed both into one script?
- Could you please also provide chapter for `readme.md` with explanation how to use it?

**Comment by rodrigok** [comment]
@szwacz done :)

File: app/package.json [hunk 1]
```diff
@@ -4,10 +4,17 @@
   "identifier": "com.example.electron-boilerplate",
   "description": "Starter for your Electron application. Out of the box ready for serious stuff.",
   "version": "0.1.0",
+  "build": "1",
```
Comment (szwacz) [inline app/package.json]: Could you please explain why this `build` property is needed? Why ordinary version can't be used?

Comment (engelgabriel) [inline app/package.json]: I believe you have to increment this on every build you submit for review. You may have multiple attempts to get a specific version approved.

Comment (rodrigok) [inline app/package.json]: Exactly

**Comment by szwacz** [comment]
Thanks. I'm going to play with this topic tomorrow, to develop some understanding of the whole process. Does anyone know how many stages of this code signing process I can do without paying Apple money?

**Comment by rodrigok** [comment]
@szwacz we are trying to approval our app, so we are doing some changes in this code, if you can wait I will update this PR with some fixes for MAS.

About your question, You need the certificates from Apple to sign the app, so you can't do anything without an apple developer subscription. :(

**Comment by szwacz** [comment]
@rodrigok as I'm looking at https://github.com/electron-userland/electron-packager it has mac app store support. Wouldn't it be simpler to incorporate that tool here?

**Comment by rodrigok** [comment]
@szwacz I have no idea :)

**Comment by rodrigok** [comment]
Hey @szwacz, our app was approved by Apple, :smile_cat: 

Can you merge this?

**Comment by szwacz** [comment]
Sure. Promise to look into it within few days.

File: tasks/release/osx.js [hunk 2]
```diff
@@ -78,21 +81,65 @@ var renameApp = function () {
...
+            'codesign --deep -f -s "' + MASIdentity + '" --entitlements child.plist -v "' + finalAppDir.path() + '/Contents/Frameworks/Electron Framework.framework/Versions/A/Libraries/libnode.dylib"',
+            'codesign --deep -f -s "' + MASIdentity + '" --entitlements child.plist -v "' + finalAppDir.path() + '/Contents/Frameworks/Electron Framework.framework/Versions/A"',
+            'codesign --deep -f -s "' + MASIdentity + '" --entitlements child.plist -v "' + finalAppDir.path() + '/Contents/Frameworks/' + manifest.productName + ' Helper.app/"',
+            'codesign --deep -f -s "' + MASIdentity + '" --entitlements child.plist -v "' + finalAppDir.path() + '/Contents/Frameworks/' + manifest.productName + ' Helper EH.app/"',
+            'codesign --deep -f -s "' + MASIdentity + '" --entitlements child.plist -v "' + finalAppDir.path() + '/Contents/Frameworks/' + manifest.productName + ' Helper NP.app/"'
+        ];
+
+        if (finalAppDir.exists('Contents/Frameworks/Squirrel.framework/Versions/A')) {
+            // # Signing a non-MAS build.
```
Comment (szwacz) [inline tasks/release/osx.js]: What is a non-MAS build if MASIdentity is used here?

File: child.plist [hunk 3]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (szwacz) [inline child.plist]: Could this file be moved to `resources/osx`?

File: parent.plist [hunk 4]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (szwacz) [inline parent.plist]: Could this file be moved to `resources/osx`?

File: tasks/utils.js [hunk 5]
```diff
@@ -27,8 +27,20 @@ module.exports.getEnvName = function () {
     return argv.env || 'development';
 };
 
-module.exports.getSigningId = function () {
-    return argv.sign;
+module.exports.getSigningId = function (manifest) {
+    return argv.sign || (manifest.codeSignIdentitiy ? manifest.codeSignIdentitiy.dmg : undefined);
```
Comment (szwacz) [inline tasks/utils.js]: So for normal app signing codeSignIdentitiy can be used as well?

File: README.md [hunk 6]
```diff
@@ -142,10 +142,42 @@ You can create Windows installer only when running on Windows, the same is true
...
+```
+
+To sign your app for Mac App Store
+```shell
+npm run release -- --mas --mas-sign "3rd Party Mac Developer Application: Company Name (APPIDENTITY)" --mas-installer-sign "3rd Party Mac Developer Installer: Company Name (APPIDENTITY)"
+```
+
+Or edit the `app/package.json`, remove the `//` from `//codeSignIdentitiy` and update the values with your sign indentities
+```json
+  "//codeSignIdentitiy": {
```
Comment (szwacz) [inline README.md]: I assume it's safe to store those strings in repo. Correct?

File: tasks/release/osx.js (hunk 2 above)
Comment (tyv) [inline tasks/release/osx.js]: you can sign app and don't want it to be in MAS

Comment (tyv) [inline tasks/release/osx.js]: btw there is already abstracted npm module for signing app: https://www.npmjs.com/package/electron-osx-sign
can be used like https://github.com/ubergrape/grape-electron/blob/master/tasks/release/osx.js#L96

File: README.md (hunk 6 above)
Comment (rodrigok) [inline README.md]: Yes, these are only the ID of the certificates, but users can pass that ids via command line too

File: child.plist (hunk 3 above)
Comment (rodrigok) [inline child.plist]: Probably yes, I will do that ASAP

File: tasks/release/osx.js (hunk 2 above)
Comment (rodrigok) [inline tasks/release/osx.js]: Idk, I copied this from an example

File: tasks/utils.js (hunk 5 above)
Comment (rodrigok) [inline tasks/utils.js]: Yes, there are 3 signs inside the `codeSignIdentitiy`, `dmg`, `MAS` and `MASInstaller`, the `dmg` is the normal app signing.

**Comment by rodrigok** [comment]
@szwacz the PR is up to date, and I did the requested changes.

**Comment by szwacz** [comment]
Thanks @rodrigok!

**Comment by tyv** [comment]
sorry guys to bring a mess to party, but you probably want to write update to doc, because of atom/electron/issues#3871 so this sign right now may brake an app and bring some time to find out the reason. So for now, signing apps at `0.37.x` can brake an app. 

I will suggest PR to doc now

**Comment by szwacz** [comment]
I've introduced one change: https://github.com/szwacz/electron-boilerplate/commit/1e83cf564d6aa2558fc320437f4f104ab9d93ca1
`pacakge.json` started to look messy, so I put all OSX relevant stuff into separate namespace.

**Pull request merged**